File: cz_cli/commands/schema.py
"""clickzetta schema command — manage schemas."""

from __future__ import annotations

from typing import Any

import click

from cz_cli import output
from cz_cli.cli_group import CLIGroup
from cz_cli.connection import get_connection
from cz_cli.connection_ctx import connection_kwargs_from_ctx
from cz_cli.logger import log_operation


@click.group("schema", cls=CLIGroup)
@click.pass_context
def schema_cmd(ctx: click.Context) -> None:
    """Manage schemas."""


@schema_cmd.command("list")
@click.option("--like", help="Filter schemas by pattern (e.g. 'test%').")
@click.option("--limit", default=100, help="Maximum number of schemas to return (default: 100).")
@click.pass_context
def list_schemas(ctx: click.Context, like: str | None, limit: int) -> None:
    """List all schemas in the current workspace."""
    fmt: str = ctx.obj.get("format", "json")
    profile: str | None = ctx.obj.get("profile")
    jdbc_url: str | None = ctx.obj.get("jdbc_url")

    try:
        conn = get_connection(jdbc_url=jdbc_url, profile=profile, **connection_kwargs_from_ctx(ctx))
    except Exception as exc:
        log_operation("schema list", ok=False, error_code="CONNECTION_ERROR")
        output.error("CONNECTION_ERROR", str(exc), fmt=fmt)
        return

    timer = output.Timer()
    try:
        with timer:
            cursor = conn.cursor()
            try:
                sql = "SHOW SCHEMAS"
                if like:
                    sql += f" LIKE '{like}'"
                # Note: SHOW SCHEMAS does not support LIMIT clause, we'll limit client-side

                cursor.execute(sql)
                rows = cursor.fetchall()

                # Convert tuples to dicts
                columns = [d[0] for d in cursor.description] if cursor.description else []
                rows_dict = [
                    dict(zip(columns, row)) if isinstance(row, tuple) else row for row in rows
                ]

                result = []
                for row in rows_dict:
                    result.append(
                        {
                            "name": row.get(
                                "schema_name",
                                row.get("name", row.get(columns[0], "") if columns else ""),
                            ),
                            "type": row.get("type", ""),
                        }
                    )

                # Client-side limit (SHOW SCHEMAS doesn't support LIMIT clause)
                total_count = len(result)
                if total_count > limit:
                    result = result[:limit]

                # Add AI message if results were limited
                ai_msg = None
                if total_count > limit:
                    ai_msg = f"Results limited to {limit} of {total_count} schemas. Use --limit to adjust or --like to filter."

                log_operation("schema list", ok=True, time_ms=timer.elapsed_ms)
                output.success(result, time_ms=timer.elapsed_ms, fmt=fmt, ai_message=ai_msg)
            finally:
                cursor.close()
    except Exception as exc:
        log_operation("schema list", ok=False, error_code="SQL_ERROR")
        output.error("SQL_ERROR", str(exc), fmt=fmt)
    finally:
        conn.close()


@schema_cmd.command("describe")
@click.argument("name")
@click.pass_context
def describe_schema(ctx: click.Context, name: str) -> None:
    """Show schema details including tables."""
    fmt: str = ctx.obj.get("format", "json")
    profile: str | None = ctx.obj.get("profile")
    jdbc_url: str | None = ctx.obj.get("jdbc_url")

    try:
        conn = get_connection(jdbc_url=jdbc_url, profile=profile, **connection_kwargs_from_ctx(ctx))
    except Exception as exc:
        log_operation("schema describe", ok=False, error_code="CONNECTION_ERROR")
        output.error("CONNECTION_ERROR", str(exc), fmt=fmt)
        return

    timer = output.Timer()
    try:
        with timer:
            cursor = conn.cursor()
            try:
                # Get schema info
                cursor.execute(f"SHOW SCHEMAS EXTENDED WHERE schema_name='{name}'")
                schema_rows = cursor.fetchall()

                if not schema_rows:
                    log_operation("schema describe", ok=False, error_code="SCHEMA_NOT_FOUND")
                    output.error("SCHEMA_NOT_FOUND", f"Schema '{name}' not found", fmt=fmt)
                    return

                # Convert tuple to dict
                columns = [d[0] for d in cursor.description] if cursor.description else []
                schema_info = (
                    dict(zip(columns, schema_rows[0]))
                    if isinstance(schema_rows[0], tuple)
                    else schema_rows[0]
                )

                # Get tables in schema
                cursor.execute(f"SHOW TABLES IN {name}")
                table_rows = cursor.fetchall()

                # Convert tuples to dicts
                table_columns = [d[0] for d in cursor.description] if cursor.description else []
                tables = []
                for r in table_rows:
                    if isinstance(r, tuple):
                        row_dict = dict(zip(table_columns, r))
                        tables.append(row_dict.get(table_columns[0], "") if table_columns else "")
                    elif isinstance(r, dict):
                        tables.append(list(r.values())[0] if r else "")
                    else:
                        tables.append(str(r))

                result: dict[str, Any] = {
                    "name": name,
                    "type": schema_info.get("type", ""),
                    "table_count": len(tables),
                    "tables": tables,
                }

                log_operation("schema describe", ok=True, time_ms=timer.elapsed_ms)
                output.success(result, time_ms=timer.elapsed_ms, fmt=fmt)
            finally:
                cursor.close()
    except Exception as exc:
        log_operation("schema describe", ok=False, error_code="SQL_ERROR")
        output.error("SQL_ERROR", str(exc), fmt=fmt)
    finally:
        conn.close()


@schema_cmd.command("create")
@click.argument("name")
@click.pass_context
def create_schema(ctx: click.Context, name: str) -> None:
    """Create a new schema."""
    fmt: str = ctx.obj.get("format", "json")
    profile: str | None = ctx.obj.get("profile")
    jdbc_url: str | None = ctx.obj.get("jdbc_url")

    try:
        conn = get_connection(jdbc_url=jdbc_url, profile=profile, **connection_kwargs_from_ctx(ctx))
    except Exception as exc:
        log_operation("schema create", ok=False, error_code="CONNECTION_ERROR")
        output.error("CONNECTION_ERROR", str(exc), fmt=fmt)
        return

    timer = output.Timer()
    try:
        with timer:
            cursor = conn.cursor()
            try:
                cursor.execute(f"CREATE SCHEMA {name}")
                log_operation("schema create", ok=True, time_ms=timer.elapsed_ms)
                output.success(
                    {"message": f"Schema '{name}' created successfully"},
                    time_ms=timer.elapsed_ms,
                    fmt=fmt,
                )
            finally:
                cursor.close()
    except Exception as exc:
        log_operation("schema create", ok=False, error_code="SQL_ERROR")
        output.error("SQL_ERROR", str(exc), fmt=fmt)
    finally:
        conn.close()


@schema_cmd.command("drop")
@click.argument("name")
@click.pass_context
def drop_schema(ctx: click.Context, name: str) -> None:
    """Drop a schema."""
    fmt: str = ctx.obj.get("format", "json")
    profile: str | None = ctx.obj.get("profile")
    jdbc_url: str | None = ctx.obj.get("jdbc_url")

    try:
        conn = get_connection(jdbc_url=jdbc_url, profile=profile, **connection_kwargs_from_ctx(ctx))
    except Exception as exc:
        log_operation("schema drop", ok=False, error_code="CONNECTION_ERROR")
        output.error("CONNECTION_ERROR", str(exc), fmt=fmt)
        return

    timer = output.Timer()
    try:
        with timer:
            cursor = conn.cursor()
            try:
                cursor.execute(f"DROP SCHEMA {name}")
                log_operation("schema drop", ok=True, time_ms=timer.elapsed_ms)
                output.success(
                    {"message": f"Schema '{name}' dropped successfully"},
                    time_ms=timer.elapsed_ms,
                    fmt=fmt,
                )
            finally:
                cursor.close()
    except Exception as exc:
        log_operation("schema drop", ok=False, error_code="SQL_ERROR")
        output.error("SQL_ERROR", str(exc), fmt=fmt)
    finally:
        conn.close()

File: packages/cz-cli/src/commands/schema.ts
import type { Argv } from "yargs"
import { commandGroup } from "../command-group.js"
import { JobStatus } from "@clickzetta/sdk"
import type { GlobalArgs } from "../cli.js"
import { success, error } from "../output/index.js"
import { logOperation } from "../logger.js"
import { getExecContext, execSql, isQueryResult, validateIdentifier, classifyExecError, rowsToRecords } from "./exec.js"
import { getShowTableName } from "./table.js"

const DEFAULT_LIMIT = 100

export function registerSchemaCommand(cli: Argv<GlobalArgs>): void {
  cli.command("schema", "Manage schemas", (yargs) => {
    yargs
      .command(
        "list",
        "List schemas",
        (y) =>
          y
            .option("like", { type: "string", describe: "Filter pattern" })
            .option("limit", { type: "number", default: DEFAULT_LIMIT, describe: "Max rows" }),
        async (argv) => {
          const format = argv.format
          try {
            const ctx = await getExecContext(argv)
            let sql = "SHOW SCHEMAS"
            if (argv.like) sql += ` LIKE '${argv.like.replace(/'/g, "''")}'`
            const limit = argv.limit ?? DEFAULT_LIMIT
            const t0 = Date.now()
            const r = await execSql(ctx, sql)
            if (!isQueryResult(r) || r.status === JobStatus.FAILED) {
              const msg = isQueryResult(r) ? (r.errorMessage ?? "Query failed") : "Unexpected result"
              logOperation("schema list", { sql, ok: false, timeMs: Date.now() - t0 })
              error(isQueryResult(r) ? (r.errorCode ?? "SQL_ERROR") : "SQL_ERROR", msg, { format }); return
            }
            let aiMessage: string | undefined
            let rows = r.rows
            if (rows.length > limit) {
              rows = rows.slice(0, limit)
              aiMessage = `Results limited to ${limit} of ${r.rows.length} schemas. Use --limit to adjust or --like to filter.`
            }
            const normalized = rows.map((row) => ({
              name: row[0] ?? "",
              type: row[1] ?? "",
            }))
            logOperation("schema list", { sql, ok: true, rows: normalized.length, timeMs: Date.now() - t0 })
            success(normalized, { format, timeMs: Date.now() - t0, aiMessage })
          } catch (err) {
            const { code: _ec, message: _em, aiMessage: _ea } = classifyExecError(err)
            error(_ec, _em, { format , ...(_ea && { aiMessage: _ea }) }); return
          }
        },
      )
      .command(
        "describe <name>",
        "Describe a schema",
        (y) => y.positional("name", { type: "string", demandOption: true, describe: "Schema name" }),
        async (argv) => {
          const format = argv.format
          try {
            const ctx = await getExecContext(argv)
            const name = validateIdentifier(argv.name as string, "schema name")
            const t0 = Date.now()
            const infoSql = `SHOW SCHEMAS EXTENDED WHERE schema_name='${name.replace(/'/g, "''")}'`
            const tablesSql = `SHOW TABLES IN ${name}`
            const [infoR, tablesR] = await Promise.all([execSql(ctx, infoSql), execSql(ctx, tablesSql)])
            const infoRows = isQueryResult(infoR) && infoR.status === JobStatus.SUCCEEDED ? infoR.rows : []
            if (infoRows.length === 0) {
              logOperation("schema describe", { sql: infoSql, ok: false, timeMs: Date.now() - t0 })
              error("SCHEMA_NOT_FOUND", `Schema '${name}' not found`, { format }); return
            }
            const schemaType = infoRows.length > 0 ? (infoRows[0][1] ?? "") : ""
            const tableResult = isQueryResult(tablesR) && tablesR.status === JobStatus.SUCCEEDED ? tablesR : undefined
            const tableRows = tableResult?.rows ?? []
            const tableRecords = tableResult ? rowsToRecords(tableResult) : []
            const tables = tableRows.map((row, index) => getShowTableName(row, tableRecords[index] ?? {}))
            logOperation("schema describe", { sql: infoSql, ok: true, timeMs: Date.now() - t0 })
            success({ name, type: schemaType, table_count: tables.length, tables }, { format, timeMs: Date.now() - t0 })
          } catch (err) {
            const { code: _ec, message: _em, aiMessage: _ea } = classifyExecError(err)
            error(_ec, _em, { format , ...(_ea && { aiMessage: _ea }) }); return
          }
        },
      )
      .command(
        "create <name>",
        "Create a schema",
        (y) => y.positional("name", { type: "string", demandOption: true, describe: "Schema name" }),
        async (argv) => {
          const format = argv.format
          try {
            const ctx = await getExecContext(argv)
            const name = validateIdentifier(argv.name as string, "schema name")
            const sql = `CREATE SCHEMA ${name}`
            const t0 = Date.now()
            const r = await execSql(ctx, sql)
            if (!isQueryResult(r) || r.status === JobStatus.FAILED) {
              const msg = isQueryResult(r) ? (r.errorMessage ?? "Query failed") : "Unexpected async result"
              logOperation("schema create", { sql, ok: false, timeMs: Date.now() - t0 })
              error(isQueryResult(r) ? (r.errorCode ?? "SQL_ERROR") : "SQL_ERROR", msg, { format }); return
            }
            logOperation("schema create", { sql, ok: true, timeMs: Date.now() - t0 })
            success({ message: `Schema '${argv.name}' created successfully` }, { format, timeMs: Date.now() - t0 })
          } catch (err) {
            const { code: _ec, message: _em, aiMessage: _ea } = classifyExecError(err)
            error(_ec, _em, { format , ...(_ea && { aiMessage: _ea }) }); return
          }
        },
      )
      .command(
        "drop <name>",
        "Drop a schema",
        (y) => y.positional("name", { type: "string", demandOption: true, describe: "Schema name" }),
        async (argv) => {
          const format = argv.format
          try {
            const ctx = await getExecContext(argv)
            const name = validateIdentifier(argv.name as string, "schema name")
            const sql = `DROP SCHEMA ${name}`
            const t0 = Date.now()
            const r = await execSql(ctx, sql)
            if (!isQueryResult(r) || r.status === JobStatus.FAILED) {
              const msg = isQueryResult(r) ? (r.errorMessage ?? "Query failed") : "Unexpected async result"
              logOperation("schema drop", { sql, ok: false, timeMs: Date.now() - t0 })
              error(isQueryResult(r) ? (r.errorCode ?? "SQL_ERROR") : "SQL_ERROR", msg, { format }); return
            }
            logOperation("schema drop", { sql, ok: true, timeMs: Date.now() - t0 })
            success({ message: `Schema '${argv.name}' dropped successfully` }, { format, timeMs: Date.now() - t0 })
          } catch (err) {
            const { code: _ec, message: _em, aiMessage: _ea } = classifyExecError(err)
            error(_ec, _em, { format , ...(_ea && { aiMessage: _ea }) }); return
          }
        },
      )
    return commandGroup(yargs, "schema")
  })
}
